### DISCUSSION TITLE
Sorting algorithm for push_swap

### DISCUSSION THREAD

**Discussion by SirAlabar** [community/community#138118] [Programming Help] [answered]
### Body

What is the most efficient way to implement the sorting algorithm for push_swap when dealing with only 3 numbers in C, considering I'm using a doubly linked list?

### Guidelines

- [X] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by MarkosComK** [comment, marked as answer]
To get a better understanding of how to optimize moves in **Push_swap**, I highly recommend this guide: [Push_swap: The Least Amount of Moves with Two Stacks](https://medium.com/@jamierobertdawson/push-swap-the-least-amount-of-moves-with-two-stacks-d1e76a71789a), which explains strategies like minimizing the number of operations and working efficiently with two stacks.

Additionally, I have a complete, well-documented solution available in my repository: [42-Push_swap](https://github.com/MarkosComK/42-Push_swap), which covers various algorithmic techniques. Below are a couple of code snippets to help you get started.

### Example of Push-Swap Operations
Here’s how the basic operations for two stacks (stack A and stack B) are implemented:

```c
// Swap the first two elements at the top of stack A
void sa(t_stack *stack)
{
    if (stack->size > 1)
    {
        int temp = stack->array[0];
        stack->array[0] = stack->array[1];
        stack->array[1] = temp;
    }
}

// Push the top element of stack B to stack A
void pa(t_stack *stack_a, t_stack *stack_b)
{
    if (stack_b->size > 0)
    {
        // Shift elements in stack A to make room for the top of stack B
        for (int i = stack_a->size; i > 0; i--)
            stack_a->array[i] = stack_a->array[i - 1];
        stack_a->array[0] = stack_b->array[0];
        // Shift elements in stack B down
        for (int i = 0; i < stack_b->size - 1; i++)
            stack_b->array[i] = stack_b->array[i + 1];
        stack_a->size++;
        stack_b->size--;
    }
}
```

### Example of Efficient Move Counting
You can minimize the number of moves by keeping track of rotations and reversals more efficiently. Here’s a basic setup:

```c
int calculate_moves(int position, int stack_size)
{
    if (position <= stack_size / 2)
        return position; // Moving up is cheaper
    else
        return stack_size - position; // Moving down is cheaper
}
```

This function helps in determining whether it's faster to rotate upwards or downwards based on the element's position in the stack.

By using these approaches and studying the provided resources, you'll be able to implement more efficient algorithms that reduce the number of moves needed to sort your stack. Feel free to explore the full solution in my repo to see these snippets in action!
